include/dali/core/fast_div.h

```diff
@@ -0,0 +1,248 @@
+    lohi<uint64_t> dif = num - den;  // this serves both as difference and comparison
+    if (static_cast<int64_t>(dif.hi) >= 0) {
+      num = dif;
+      q |= static_cast<uint64_t>(1) << sh;
+    }
+    sh--;
+    den >>= 1;
+  }
+  return q;
+#endif
+}
+
+}  // namespace detail
+
+/**
+ * @brief Fast unsigned integer division
+ *
+ * Based on:
+ * Labor of Division (Episode III): Faster Unsigned Division by Constants
+ * ridiculous_fish corydoras@ridiculousfish.com
+ * https://ridiculousfish.com/blog/posts/labor-of-division-episode-iii.html
+ *
+ * Limitations:
+ * - doesn't work for maximum value of uint when dividing by uncooperative numbers
+ */
+template <typename uint>
+struct fast_div {
+  uint divisor;
+  uint mul;
+  uint8_t add;
+  uint8_t shift;
+
+  DALI_HOST_DEV fast_div() {}
+
+  DALI_HOST_DEV fast_div(uint divisor) {  // NOLINT
+    init(divisor);
+  }
+
+  DALI_HOST_DEV void init(uint divisor) {
+    this->divisor = divisor;
+    this->mul = 1;
+    this->shift = 0;
+    this->add = 0;
+    if (divisor == 0) {
+      return;
+    }
+    int l = ilog2(divisor);
+
+    this->shift = l;
+
+    if ((divisor & (divisor - 1)) == 0) {
+      this->mul = 0;
+      this->shift = l;
+      return;
+    }
+
+    uint m_lo = detail::div_lohi(0,            uint(1) << l, divisor);
+    uint m_hi = detail::div_lohi(uint(1) << l, uint(1) << l, divisor);
+    this->add = (m_lo == m_hi);  // round-up failed, use round-down method
+    this->mul = m_hi;
+  }
+
+  DALI_HOST_DEV DALI_FORCEINLINE operator uint() const noexcept {
+    return divisor;
+  }
+};
+
+DALI_HOST_DEV DALI_FORCEINLINE uint32_t operator/(uint32_t x, fast_div<uint32_t> y) {
+  // If the divisor is a power of 2, the multiplier would be 2^32, which is out of range
+  // - therefore, powers of 2 get special treatment and the multiplication is skipped.
+#ifdef __CUDA_ARCH__
+  x = y.mul ? __umulhi(x + y.add, y.mul) : x;
+  return x >> y.shift;
+#else
+  if (y.mul) {
+    uint32_t hi = static_cast<uint64_t>(x + y.add) * y.mul >> 32;
+    return hi >> y.shift;
+  } else {
+    return x >> y.shift;
+  }
+#endif
+}
+
+DALI_HOST_DEV DALI_FORCEINLINE uint64_t operator/(uint64_t x, fast_div<uint64_t> y) {
+  // If the divisor is a power of 2, the multiplier would be 2^64, which is out of range
+  // - therefore, powers of 2 get special treatment and the multiplication is skipped.
+#ifdef __CUDA_ARCH__
+  x = y.mul ? __umul64hi(x + y.add, y.mul) : x;
+  return x >> y.shift;
+#else
+  if (y.mul) {
+    uint64_t hi = static_cast<unsigned __int128>(x + y.add) * y.mul >> 64;
+    return hi >> y.shift;
+  } else {
+    return x >> y.shift;
+  }
+#endif
+}
+
+template <typename uint>
+DALI_HOST_DEV DALI_FORCEINLINE uint operator%(uint x, fast_div<uint> y) {
+  return x - x / y * y;
+}
+
+template <typename uint>
+DALI_HOST_DEV DALI_FORCEINLINE uint div_mod(uint &mod, uint dividend, fast_div<uint> divisor) {
+  uint q = dividend / divisor;
+  mod = dividend - q * divisor;
+  return q;
+}
```

Comment:
How about this kind of API?
```c++
template <typename uint>
std::pair<uint, uint> div_mod(uint dividend, fast_div<uint> divisor) {
  uint q = dividend / divisor;
  return {q, dividend - q * divisor};
}
```

Reply:
Won't work in `__device__` code - std::pair constructor is a `__host __` function.